staging/src/k8s.io/api/events/v1/types.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Event is a report of an event somewhere in the cluster. It generally denotes some state change in the system.
+type Event struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Required. Time when this Event was first observed.
+	EventTime metav1.MicroTime `json:"eventTime" protobuf:"bytes,2,opt,name=eventTime"`
```

Comment:
looking at ValidateEvent, it's unclear to me how the validation actually works... it looks like:

if this field is zero-value:
* only the namespace of the event and of involvedObject is verified to match
* none of the other fields are validated

if this field is non-zero-value:
* the involved object namespace is allowed to mismatch

There's the following comment in ValidateEvent:
```go
// "New" Events need to have EventTime set, so it's validating old object.
```
but I don't see where that is enforced (the event strategy and defaulting doesn't populate EventTime that I can see, and the strategy and validation don't seem to prevent clearing EventTime on update of an existing event)

Reply:
@wojtek-t @yastij 
I'm not sure how Strategy is being used when creating and updating Events. To me it seems only to check fields based on EventTime availability, but doesn't do anything with EventTime field. Any thoughts about how to improve this?

Reply:
the if/else has a comment that the looser validation is for compatibility with existing clients. API version boundaries are the right time to start enforcing validation consistently via the new API. As an example, you can see how we validate CertificateSigningRequests more strictly via the v1 API (https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/certificates/certificates/strategy.go#L96-L100)

Reply:
I agree with Jordan. And it shouldn't be hard to separate that logic - it's already well splited between old and new event.

Reply:
The new validation was added - I'm resolving this comment. We can continue discussing on the new validation file.